From 5240c3e8aa1b183041568c3cd85d348163586d37 Mon Sep 17 00:00:00 2001
From: user <user@computer>
Date: Sat, 19 Sep 2020 14:42:00 +0200
Subject: [PATCH] mwlwifi: use latest firmware and disable power/region lock

This patch does two things:
1. forward the driver's revision to the latest one at the time of this patch
   This enables a newer firmware to be used.
2. Don't enforce the region and power table from EEPROM for MWL8964
   This doesn't actually allow you to override the EEPROM settings! It just
   ignores mismatches between EEPROM and OpenWRT setting and thus prevents
   the driver from refusing to load on mismatch. In addition, it allows DFS
   channels to be used on non-US versions of WRT32X and WRT3200ACM, which are
   otherwise blocked.
---
 package/kernel/mwlwifi/Makefile               |   6 +-
 ...1-remove-region-and-power-table-lock.patch | 209 ++++++++++++++++++
 2 files changed, 212 insertions(+), 3 deletions(-)
 create mode 100644 package/kernel/mwlwifi/patches/001-remove-region-and-power-table-lock.patch

diff --git a/package/kernel/mwlwifi/Makefile b/package/kernel/mwlwifi/Makefile
index 642d800836..26fc845362 100644
--- a/package/kernel/mwlwifi/Makefile
+++ b/package/kernel/mwlwifi/Makefile
@@ -15,9 +15,9 @@ PKG_LICENSE_FILES:=
 
 PKG_SOURCE_URL:=https://github.com/kaloz/mwlwifi
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2019-03-02
-PKG_SOURCE_VERSION:=31d9386079b91cc699658c19294e139b62b512bc
-PKG_MIRROR_HASH:=7bdd05765d8215a9c293cdcb028d63a04c9e55b337eaac9e8d3659bd86218321
+PKG_SOURCE_DATE:=2020-04-16
+PKG_SOURCE_VERSION:=111118dc2ea3b592a5f7dff18c82d57a651970e7
+PKG_MIRROR_HASH:=dea6eeb2200ebe52a8f5db5a188b85fb6e57b68fec5f6d3437375e63bf77741b
 
 PKG_MAINTAINER:=Imre Kaloz <kaloz@openwrt.org>
 PKG_BUILD_PARALLEL:=1
diff --git a/package/kernel/mwlwifi/patches/001-remove-region-and-power-table-lock.patch b/package/kernel/mwlwifi/patches/001-remove-region-and-power-table-lock.patch
new file mode 100644
index 0000000000..f25b8300d8
--- /dev/null
+++ b/package/kernel/mwlwifi/patches/001-remove-region-and-power-table-lock.patch
@@ -0,0 +1,209 @@
+Remove locked region and power table for WRT3200ACM and WRT32X
+
+--- a/core.c
++++ b/core.c
+@@ -385,13 +385,6 @@ static void mwl_regd_init(struct mwl_priv *priv)
+ 	/* hook regulatory domain change notification */
+ 	priv->hw->wiphy->reg_notifier = mwl_reg_notifier;
+ 
+-	if (priv->chip_type == MWL8964)
+-		rc = mwl_fwcmd_get_pwr_tbl_sc4(priv->hw,
+-					       &priv->device_pwr_tbl[0],
+-					       &region_code,
+-					       &priv->number_of_channels,
+-					       0);
+-	else
+ 		rc = mwl_fwcmd_get_device_pwr_tbl(priv->hw,
+ 						  &priv->device_pwr_tbl[0],
+ 						  &region_code,
+@@ -403,13 +396,6 @@ static void mwl_regd_init(struct mwl_priv *priv)
+ 	priv->forbidden_setting = true;
+ 
+ 	for (i = 1; i < priv->number_of_channels; i++) {
+-		if (priv->chip_type == MWL8964)
+-			mwl_fwcmd_get_pwr_tbl_sc4(priv->hw,
+-						  &priv->device_pwr_tbl[i],
+-						  &region_code,
+-						  &priv->number_of_channels,
+-						  i);
+-		else
+ 			mwl_fwcmd_get_device_pwr_tbl(priv->hw,
+ 						     &priv->device_pwr_tbl[i],
+ 						     &region_code,
+@@ -858,10 +844,6 @@ static int mwl_wl_init(struct mwl_priv *priv)
+ 		mwl_fwcmd_get_txpwrlmt_cfg_data(hw);
+ 	}
+ 
+-	if (priv->chip_type == MWL8964)
+-		rc = mwl_fwcmd_get_fw_region_code_sc4(hw,
+-						      &priv->fw_region_code);
+-	else
+ 		rc = mwl_fwcmd_get_fw_region_code(hw, &priv->fw_region_code);
+ 	if (!rc) {
+ 		priv->fw_device_pwrtbl = true;
+--- a/hif/fwcmd.c
++++ b/hif/fwcmd.c
+@@ -94,8 +94,6 @@ char *mwl_fwcmd_get_cmd_string(unsigned short cmd)
+ 		{ HOSTCMD_CMD_GET_DEVICE_PWR_TBL, "GetDevicePwrTbl" },
+ 		{ HOSTCMD_CMD_SET_RATE_DROP, "SetRateDrop" },
+ 		{ HOSTCMD_CMD_NEWDP_DMATHREAD_START, "NewdpDMAThreadStart" },
+-		{ HOSTCMD_CMD_GET_FW_REGION_CODE_SC4, "GetFwRegionCodeSC4" },
+-		{ HOSTCMD_CMD_GET_DEVICE_PWR_TBL_SC4, "GetDevicePwrTblSC4" },
+ 		{ HOSTCMD_CMD_QUIET_MODE, "QuietMode" },
+ 		{ HOSTCMD_CMD_CORE_DUMP_DIAG_MODE, "CoreDumpDiagMode" },
+ 		{ HOSTCMD_CMD_802_11_SLOT_TIME, "80211SlotTime" },
+@@ -3431,83 +3429,7 @@ int mwl_fwcmd_newdp_dmathread_start(struct ieee80211_hw *hw)
+ }
+ 
+ 
+-int mwl_fwcmd_get_fw_region_code_sc4(struct ieee80211_hw *hw,
+-				     u32 *fw_region_code)
+-{
+-	struct mwl_priv *priv = hw->priv;
+-	struct hostcmd_cmd_get_fw_region_code_sc4 *pcmd;
+-	u16 cmd;
+-
+-	pcmd = (struct hostcmd_cmd_get_fw_region_code_sc4 *)&priv->pcmd_buf[0];
+-
+-	mutex_lock(&priv->fwcmd_mutex);
+-
+-	memset(pcmd, 0x00, sizeof(*pcmd));
+-	cmd = HOSTCMD_CMD_GET_FW_REGION_CODE_SC4;
+-	pcmd->cmd_hdr.cmd = cpu_to_le16(cmd);
+-	pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+-
+-	if (mwl_hif_exec_cmd(hw, cmd)) {
+-		mutex_unlock(&priv->fwcmd_mutex);
+-		return -EIO;
+-	}
+-
+-	if (pcmd->cmd_hdr.result != 0) {
+-		mutex_unlock(&priv->fwcmd_mutex);
+-		return -EINVAL;
+-	}
+-
+-	if (pcmd->status)
+-		*fw_region_code = (pcmd->status == 1) ? 0 : pcmd->status;
+-	else
+-		*fw_region_code = le32_to_cpu(pcmd->fw_region_code);
+-
+-	mutex_unlock(&priv->fwcmd_mutex);
+-
+-	return 0;
+-}
+-
+-int mwl_fwcmd_get_pwr_tbl_sc4(struct ieee80211_hw *hw,
+-			      struct mwl_device_pwr_tbl *device_ch_pwrtbl,
+-			      u8 *region_code,
+-			      u8 *number_of_channels,
+-			      u32 channel_index)
+-{
+-	struct mwl_priv *priv = hw->priv;
+-	struct hostcmd_cmd_get_device_pwr_tbl_sc4 *pcmd;
+-	int status;
+-	u16 cmd;
+-
+-	pcmd = (struct hostcmd_cmd_get_device_pwr_tbl_sc4 *)&priv->pcmd_buf[0];
+-
+-	mutex_lock(&priv->fwcmd_mutex);
+-
+-	memset(pcmd, 0x00, sizeof(*pcmd));
+-	cmd = HOSTCMD_CMD_GET_DEVICE_PWR_TBL_SC4;
+-	pcmd->cmd_hdr.cmd = cpu_to_le16(cmd);
+-	pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+-	pcmd->status = cpu_to_le16(cmd);
+-	pcmd->current_channel_index = cpu_to_le32(channel_index);
+-
+-	if (mwl_hif_exec_cmd(hw, cmd)) {
+-		mutex_unlock(&priv->fwcmd_mutex);
+-		return -EIO;
+-	}
+ 
+-	device_ch_pwrtbl->channel = pcmd->channel_pwr_tbl.channel;
+-	memcpy(device_ch_pwrtbl->tx_pwr, pcmd->channel_pwr_tbl.tx_pwr,
+-	       SYSADPT_TX_PWR_LEVEL_TOTAL_SC4);
+-	device_ch_pwrtbl->dfs_capable = pcmd->channel_pwr_tbl.dfs_capable;
+-	device_ch_pwrtbl->ax_ant = pcmd->channel_pwr_tbl.ax_ant;
+-	device_ch_pwrtbl->cdd = pcmd->channel_pwr_tbl.cdd;
+-	*region_code = pcmd->region_code;
+-	*number_of_channels = pcmd->number_of_channels;
+-	status = le16_to_cpu(pcmd->status);
+-
+-	mutex_unlock(&priv->fwcmd_mutex);
+-
+-	return status;
+-}
+ 
+ int mwl_fwcmd_quiet_mode(struct ieee80211_hw *hw, bool enable, u32 period,
+ 			 u32 duration, u32 next_offset)
+--- a/hif/fwcmd.h
++++ b/hif/fwcmd.h
+@@ -253,15 +253,6 @@ int mwl_fwcmd_set_rate_drop(struct ieee80211_hw *hw, int enable,
+ 
+ int mwl_fwcmd_newdp_dmathread_start(struct ieee80211_hw *hw);
+ 
+-int mwl_fwcmd_get_fw_region_code_sc4(struct ieee80211_hw *hw,
+-				     u32 *fw_region_code);
+-
+-int mwl_fwcmd_get_pwr_tbl_sc4(struct ieee80211_hw *hw,
+-			      struct mwl_device_pwr_tbl *device_ch_pwrtbl,
+-			      u8 *region_code,
+-			      u8 *number_of_channels,
+-			      u32 channel_index);
+-
+ int mwl_fwcmd_quiet_mode(struct ieee80211_hw *hw, bool enable, u32 period,
+ 			 u32 duration, u32 next_offset);
+ 
+--- a/hif/hostcmd.h
++++ b/hif/hostcmd.h
+@@ -73,8 +73,6 @@
+ #define HOSTCMD_CMD_GET_DEVICE_PWR_TBL          0x116B
+ #define HOSTCMD_CMD_SET_RATE_DROP               0x1172
+ #define HOSTCMD_CMD_NEWDP_DMATHREAD_START       0x1189
+-#define HOSTCMD_CMD_GET_FW_REGION_CODE_SC4      0x118A
+-#define HOSTCMD_CMD_GET_DEVICE_PWR_TBL_SC4      0x118B
+ #define HOSTCMD_CMD_QUIET_MODE                  0x1201
+ #define HOSTCMD_CMD_CORE_DUMP_DIAG_MODE         0x1202
+ #define HOSTCMD_CMD_802_11_SLOT_TIME            0x1203
+@@ -1110,41 +1108,6 @@ struct hostcmd_cmd_newdp_dmathread_start {
+ 	struct hostcmd_header cmd_hdr;
+ } __packed;
+ 
+-/* HOSTCMD_CMD_GET_FW_REGION_CODE_SC4 */
+-struct hostcmd_cmd_get_fw_region_code_sc4 {
+-	struct hostcmd_header cmd_hdr;
+-	__le32 status; /* 0 = Found, 1 = Error */
+-	__le32 fw_region_code;
+-} __packed;
+-
+-/* HOSTCMD_CMD_GET_DEVICE_PWR_TBL_SC4 */
+-#define HAL_TRPC_ID_MAX_SC4        32
+-#define MAX_GROUP_PER_CHANNEL_5G   39
+-#define MAX_GROUP_PER_CHANNEL_2G   21
+-#define	MAX(a, b) (((a) > (b)) ? (a) : (b))
+-#define MAX_GROUP_PER_CHANNEL_RATE \
+-	MAX(MAX_GROUP_PER_CHANNEL_5G, MAX_GROUP_PER_CHANNEL_2G)
+-
+-struct channel_power_tbl_sc4 {
+-	u8 channel;
+-	u8 grp_pwr[MAX_GROUP_PER_CHANNEL_RATE];
+-	u8 tx_pwr[HAL_TRPC_ID_MAX_SC4];
+-	u8 dfs_capable;
+-	u8 ax_ant;
+-	u8 cdd;
+-	u8 rsvd;
+-} __packed;
+-
+-struct hostcmd_cmd_get_device_pwr_tbl_sc4 {
+-	struct hostcmd_header cmd_hdr;
+-	__le16 status; /* 0 = Found, 1 = Error */
+-	u8 region_code;
+-	u8 number_of_channels;
+-	__le32 current_channel_index;
+-	/* Only for 1 channel, so, 1 channel at a time */
+-	struct channel_power_tbl_sc4 channel_pwr_tbl;
+-} __packed;
+-
+ /* HOSTCMD_CMD_QUIET_MODE */
+ struct hostcmd_cmd_quiet_mode {
+ 	struct hostcmd_header cmd_hdr;
-- 
2.20.1

